Tests/scripts/hook_validations/docker.py

```diff
@@ -0,0 +1,239 @@
+    """
+    res = requests.get(
+        'https://{}/v2/'.format(registry),
+        headers=ACCEPT_HEADER,
+        timeout=TIMEOUT,
+        verify=verify_ssl
+    )
+    if res.status_code == 401:  # need to authenticate
+        # defaults in case we fail for some reason
+        realm = "https://auth.docker.io/token"
+        service = "registry.docker.io"
+        # Should contain header: Www-Authenticate
+        www_auth = res.headers.get('www-authenticate')
+        if www_auth:
+            parse_auth = parse_www_auth(www_auth)
+            if parse_auth and len(parse_auth) == 2:
+                realm, service = parse_auth
+        res = requests.get(
+            '{}?scope=repository:{}:pull&service={}'.format(realm, image_name, service),
+            headers=ACCEPT_HEADER,
+            timeout=TIMEOUT,
+            verify=verify_ssl
+        )
+        res.raise_for_status()
+        res_json = res.json()
+        return res_json.get('token')
+    else:
+        res.raise_for_status()
+        return None
+
+
+def clear_non_numbered_tags(tags):
+    """Clears a given tags list to only keep numbered tags
+
+    Args:
+        tags(list): list of docker image tag names - ordered in lexical order
+
+    Returns:
+        a tag list with only numbered tags
+    """
+    only_numbered_tags = []
+    for tag in tags:
+        number_token = 1
+        split_tag = tag.split('.')
+        for sub_section in split_tag:
+            if not sub_section.isdigit():
+                number_token = 0
+
+        if number_token:
+            only_numbered_tags.append(tag)
+
+    return only_numbered_tags
+
+
+def lexical_find_latest_tag(tags):
+    """Will return the latest numeric docker image tag if possible - otherwise will return the last lexical tag.
+
+    for example for the tag list: [2.0.2000, 2.1.2700 2.1.373, latest], will return 2.1.2700
+
+    Args:
+        tags(list): list of docker image tag names - ordered in lexical order
+    """
+
+    only_numbered_tags = clear_non_numbered_tags(tags)
+
+    if len(only_numbered_tags) == 0:
+        return tags[-1]
+
+    max_tag = only_numbered_tags[0]
+
+    for num_tag in only_numbered_tags:
+        if parse_version(max_tag) < parse_version(num_tag):
+            max_tag = num_tag
+
+    return max_tag
+
+
+def find_latest_tag_by_date(tags):
+    """Get the latest tags by datetime comparison.
+
+    Args:
+        tags(list): List of dictionaries representing the docker image tags
+
+    Returns:
+        The last updated docker image tag name
+    """
+    latest_tag_name = 'latest'
+    latest_tag_date = datetime.now() - timedelta(days=400000)
+    for tag in tags:
+        tag_date = datetime.strptime(tag.get('last_updated'), '%Y-%m-%dT%H:%M:%S.%fZ')
+        if tag_date >= latest_tag_date:
+            latest_tag_date = tag_date
+            latest_tag_name = tag.get('name')
+
+    return latest_tag_name
+
+
+def get_docker_image_latest_tag(docker_image_name):
```

Comment:
is this from the script we have in Content?

Reply:
Yes, the same functions with minor changes. Mostly error printing instead of exception raising.